### PR TITLE
[pigeon] Make Linux type declarations public

### PR DESCRIPTION
**Pull request by stuartmorgan**
Moves codec and HostAPI class declarations to the header to work around a glib bug. We could instead add an unnecessary call to the type check method, but we will need the codec to be public eventually anyway, and HostApi is conceptually a public class (it just isn't strictly necessary to declare it in the header due to the way abstract interfaces work in gobject) so may as well be declared publicly.

Part of https://github.com/flutter/flutter/issues/153083

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I [linked to at least one issue that this PR fixes] in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style], or this PR is [exempt from CHANGELOG changes].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[Discord]: https://github.com/flutter/flutter/blob/master/docs/contributing/Chat.md
[linked to at least one issue that this PR fixes]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#overview
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#version
[following repository CHANGELOG style]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog-style
[exempt from CHANGELOG changes]: https://github.com/flutter/flutter/blob/master/docs/ecosystem/contributing/README.md#changelog
[test-exempt]: https://github.com/flutter/flutter/blob/master/docs/contributing/Tree-hygiene.md#tests

### CODE REVIEW

File: packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift [hunk 1]
```diff
@@ -15,24 +15,6 @@ import Foundation
...
-    self.code = code
-    self.message = message
-    self.details = details
-  }
-
-  var localizedDescription: String {
-    return
-      "PigeonError(code: \(code), message: \(message ?? "<nil>"), details: \(details ?? "<nil>")"
-  }
-}
```
[the remark this answers is not in the document]
Comment (tarrinneal) [reply packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift:34]: I thought we did too

[the remark this answers is not in the document]
Comment (stuartmorgan) [reply packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift:34]: It looks like this change breaks things, so I'll back it out. Can you look into why/how we have latent generation failures sometime soon?

[the remark this answers is not in the document]
Comment (tarrinneal) [reply packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift]: https://github.com/flutter/packages/pull/7849 no gen here. I'll take a look into why.

[the remark this answers is not in the document]
Comment (tarrinneal) [reply packages/pigeon/platform_tests/test_plugin/ios/Classes/CoreTests.gen.swift]: Are we not generating the files before testing the format? Seems like as long as the new files aren't added, it never checks them.

**Review by tarrinneal — APPROVED** [review]
Thanks!

**Pull request merged**